File: televault/lib/services/telegram/channel_service.dart
import 'dart:async';

import '../../models/caption_codec.dart';
import '../../models/vault_entry.dart';
import '../index_db.dart';
import 'td_client.dart';

const vaultMarker = '#televault-v1';

class ChannelService {
  final TdSender _td;
  final IndexDb _db;

  final changes = StreamController<void>.broadcast();
  StreamSubscription<Map<String, dynamic>>? _updatesSub;
  final _pendingUploadIds = <int>{};
  final _pendingUploadPaths = <String>{};

  ChannelService(this._td, this._db);

  void markUploadPath(String destPath) {
    if (destPath.isNotEmpty) _pendingUploadPaths.add(destPath);
  }

  void clearUploadPath(String destPath) {
    if (destPath.isNotEmpty) _pendingUploadPaths.remove(destPath);
  }

  void markUploadPending(int messageId) {
    if (messageId > 0) _pendingUploadIds.add(messageId);
  }

  void clearUploadPending(int? messageId, [int? otherId]) {
    for (final id in [messageId, otherId]) {
      if (id != null && id > 0) _pendingUploadIds.remove(id);
    }
  }

  void rekeyUploadPending(int oldId, int newId) {
    if (_pendingUploadIds.remove(oldId) && newId > 0) {
      _pendingUploadIds.add(newId);
    }
  }

  bool _shouldIndexMessage(Map<String, dynamic> msg) {
    if (!_isDocumentMessage(msg)) return true;
    final id = msg['id'];
    if (id is num && id.toInt() > 0 && _pendingUploadIds.contains(id.toInt())) {
      return false;
    }
    final entry = entryFromMessage(msg);
    if (entry != null && _pendingUploadPaths.contains(entry.path)) return false;
    return true;
  }

  Future<void> dispose() async {
    await _updatesSub?.cancel();
    _updatesSub = null;
  }

  Future<int> createVaultChannel() async {
    final chat = await _td.send({
      '@type': 'createNewSupergroupChat',
      'title': 'TeleVault Storage',
      'is_channel': true,
      'description': 'Kho file TeleVault — không xoá kênh này. $vaultMarker',
    });
    return chat['id'] as int;
  }

  Future<int?> findVaultChannel() async {
    final chats = await _td.send({'@type': 'getChats', 'limit': 200});
    for (final chatId in (chats['chat_ids'] as List? ?? const [])) {
      final chat = await _td.send({'@type': 'getChat', 'chat_id': chatId});
      final type = chat['type'] as Map<String, dynamic>?;
      if (type?['@type'] != 'chatTypeSupergroup' || type?['is_channel'] != true) continue;
      final full = await _td.send({
        '@type': 'getSupergroupFullInfo',
        'supergroup_id': type!['supergroup_id'],
      });
      if ((full['description'] as String? ?? '').contains(vaultMarker)) {
        return chatId as int;
      }
    }
    return null;
  }

  Future<bool> _isVaultChat(int chatId) async {
    try {
      final chat = await _td.send({'@type': 'getChat', 'chat_id': chatId});
      final type = chat['type'] as Map<String, dynamic>?;
      if (type?['@type'] != 'chatTypeSupergroup' || type?['is_channel'] != true) return false;
      final full = await _td.send({
        '@type': 'getSupergroupFullInfo',
        'supergroup_id': type!['supergroup_id'],
      });
      return (full['description'] as String? ?? '').contains(vaultMarker);
    } catch (_) {
      return false;
    }
  }

  /// Dùng chat id đã lưu (2 request) thay vì quét toàn bộ getChats mỗi lần đăng nhập.
  Future<int> resolveVaultChatId() async {
    final cached = await _db.getVaultChatId();
    if (cached != null && await _isVaultChat(cached)) return cached;
    final found = await findVaultChannel();
    if (found != null) {
      await _db.setVaultChatId(found);
      return found;
    }
    final created = await createVaultChannel();
    await _db.setVaultChatId(created);
    return created;
  }

  VaultEntry? entryFromMessage(Map<String, dynamic> msg) {
    final content = msg['content'] as Map<String, dynamic>?;
    final id = msg['id'] as int;
    switch (content?['@type']) {
      case 'messageDocument':
        final caption = (content!['caption'] as Map<String, dynamic>?)?['text'] as String? ?? '';
        return decodeCaption(id, caption);
      case 'messageText':
        final text = (content!['text'] as Map<String, dynamic>?)?['text'] as String? ?? '';
        return decodeCaption(id, text);
    }
    return null;
  }

  bool _isDocumentMessage(Map<String, dynamic> msg) {
    final content = msg['content'] as Map<String, dynamic>?;
    return content?['@type'] == 'messageDocument';
  }

  int? tdFileIdFromMessage(Map<String, dynamic> msg) {
    final content = msg['content'] as Map<String, dynamic>?;
    if (content?['@type'] != 'messageDocument') return null;
    final doc = content!['document'] as Map<String, dynamic>?;
    final file = doc?['document'] as Map<String, dynamic>?;
    final id = file?['id'];
    if (id is int) return id;
    if (id is num) return id.toInt();
    return null;
  }

  Future<void> scanHistory(int chatId, {void Function(int scanned)? onProgress}) async {
    var fromMessageId = 0;
    var scanned = 0;
    var maxId = await _db.getLastMessageId();
    final seenIds = <int>{};
    while (true) {
      final page = await _td.send({
        '@type': 'getChatHistory',
        'chat_id': chatId,
        'from_message_id': fromMessageId,
        'offset': 0,
        'limit': 100,
        'only_local': false,
      });
      final messages = (page['messages'] as List? ?? const []).cast<Map<String, dynamic>>();
      if (messages.isEmpty) break;
      for (final msg in messages) {
        final id = msg['id'] as int;
        seenIds.add(id);
        final entry = entryFromMessage(msg);
        if (entry != null) await _db.upsert(entry);
        scanned++;
        if (id > maxId) maxId = id;
      }
      final nextFrom = messages.last['id'] as int;
      if (nextFrom == fromMessageId) break;
      fromMessageId = nextFrom;
      onProgress?.call(scanned);
    }
    await _db.reconcileToMessageIds(seenIds);
    await _db.setLastMessageId(maxId);
    changes.add(null);
  }

  void listenUpdates(int chatId) {
    _updatesSub?.cancel();
    _updatesSub = _td.updates.listen((u) async {
      switch (u['@type']) {
        case 'updateNewMessage':
          final msg = u['message'] as Map<String, dynamic>;
          if (msg['chat_id'] != chatId) return;
          final id = msg['id'];
          if (id is num && id.toInt() < 0) return;
          if (!_shouldIndexMessage(msg)) return;
          final entry = entryFromMessage(msg);
          if (entry != null) {
            await _db.upsert(entry);
            changes.add(null);
          }
        case 'updateMessageSendSucceeded':
          final msg = u['message'] as Map<String, dynamic>?;
          if (msg == null || msg['chat_id'] != chatId) return;
          final oldId = u['old_message_id'];
          final newId = msg['id'];
          if (oldId is num && newId is num) {
            rekeyUploadPending(oldId.toInt(), newId.toInt());
          }
          if (oldId is num && newId is num && oldId.toInt() != newId.toInt()) {
            await _db.rekeyMessageId(oldId.toInt(), newId.toInt());
          }
          if (!_isDocumentMessage(msg)) {
            final entry = entryFromMessage(msg);
            if (entry != null) {
              await _db.upsert(entry);
              changes.add(null);
            }
          }
        case 'updateMessageContent':
          if (u['chat_id'] != chatId) return;
          final msgId = u['message_id'];
          if (msgId is num && _pendingUploadIds.contains(msgId.toInt())) return;
          final entry = entryFromMessage({
            'id': u['message_id'],
            'content': u['new_content'],
          });
          if (entry != null) {
            await _db.upsert(entry);
            changes.add(null);
          }
        case 'updateDeleteMessages':
          if (u['chat_id'] != chatId || u['is_permanent'] != true) return;
          for (final id in (u['message_ids'] as List? ?? const [])) {
            await _db.delete(id as int);
          }
          changes.add(null);
      }
    });
  }
}

File: electron/lib/telegram/channelService.js
const { decodeCaption } = require('@televault/core');

const VAULT_MARKER = '#televault-v1';

class ChannelService {
  /**
   * @param {import('tdl').Client} client
   * @param {import('../db/indexDb')} db
   */
  constructor(client, db) {
    this.client = client;
    this.db = db;
    /** @type {((() => void) | null)} */
    this._onChange = null;
    /** @type {((oldId: number, newId: number) => void) | null} */
    this._onMessageSendSucceeded = null;
    /** @type {((oldId: number) => void) | null} */
    this._onMessageSendFailed = null;
    /** @type {((update: unknown) => void) | null} */
    this._updateHandler = null;
    /** Message ids của upload đang chạy — không index document cho đến khi upload xong. */
    this._pendingUploadIds = new Set();
    /** Vault paths đang upload — đặt trước sendMessage để chặn updateNewMessage đồng bộ. */
    this._pendingUploadPaths = new Set();
  }

  /** @param {string} destPath */
  markUploadPath(destPath) {
    if (destPath) this._pendingUploadPaths.add(destPath);
  }

  /** @param {string} destPath */
  clearUploadPath(destPath) {
    if (destPath) this._pendingUploadPaths.delete(destPath);
  }

  /** @param {number} messageId */
  markUploadPending(messageId) {
    if (messageId > 0) this._pendingUploadIds.add(messageId);
  }

  /** @param {...(number|null|undefined)} messageIds */
  clearUploadPending(...messageIds) {
    for (const id of messageIds) {
      if (id != null && id > 0) this._pendingUploadIds.delete(id);
    }
  }

  /** @param {number} oldId @param {number} newId */
  rekeyUploadPending(oldId, newId) {
    if (this._pendingUploadIds.has(oldId)) {
      this._pendingUploadIds.delete(oldId);
      if (newId > 0) this._pendingUploadIds.add(newId);
    }
  }

  /** @param {Record<string, unknown>} msg */
  _shouldIndexMessage(msg) {
    if (!this._isDocumentMessage(msg)) return true;
    const id = this._tdId(msg.id);
    if (id > 0 && this._pendingUploadIds.has(id)) return false;
    const entry = this.entryFromMessage(msg);
    if (entry && this._pendingUploadPaths.has(entry.path)) return false;
    return true;
  }

  onChange(fn) {
    this._onChange = fn;
  }

  onMessageSendSucceeded(fn) {
    this._onMessageSendSucceeded = fn;
  }

  onMessageSendFailed(fn) {
    this._onMessageSendFailed = fn;
  }

  _tdId(v) {
    if (typeof v === 'number') return v;
    if (typeof v === 'string') return Number.parseInt(v, 10) || -1;
    return -1;
  }

  _emitChange() {
    this._onChange?.();
  }

  async createVaultChannel() {
    const chat = await this.client.invoke({
      _: 'createNewSupergroupChat',
      title: 'TeleVault Storage',
      is_channel: true,
      description: `Kho file TeleVault — không xoá kênh này. ${VAULT_MARKER}`,
    });
    return chat.id;
  }

  async findVaultChannel() {
    const chats = await this.client.invoke({
      _: 'getChats',
      chat_list: { _: 'chatListMain' },
      limit: 200,
    });
    for (const chatId of chats.chat_ids || []) {
      try {
        const chat = await this.client.invoke({ _: 'getChat', chat_id: chatId });
        const type = chat.type;
        if (type?._ !== 'chatTypeSupergroup' || !type.is_channel) continue;
        
        const supergroup = await this.client.invoke({
          _: 'getSupergroup',
          supergroup_id: type.supergroup_id,
        });
        if (supergroup.status?._ !== 'chatMemberStatusCreator') continue;

        const full = await this.client.invoke({
          _: 'getSupergroupFullInfo',
          supergroup_id: type.supergroup_id,
        });
        if (String(full.description || '').includes(VAULT_MARKER)) {
          // Network check to ensure we ACTUALLY have access
          await this.client.invoke({
            _: 'getChatHistory',
            chat_id: chatId,
            from_message_id: 0,
            offset: 0,
            limit: 1,
            only_local: false,
          });
          return chatId;
        }
      } catch (e) {
        console.warn('[TeleVault] Skipping chat during search due to error:', e.message);
      }
    }
    return null;
  }

  async isVaultChat(chatId) {
    try {
      const chat = await this.client.invoke({ _: 'getChat', chat_id: chatId });
      const type = chat.type;
      if (type?._ !== 'chatTypeSupergroup' || !type.is_channel) return false;
      
      const supergroup = await this.client.invoke({
        _: 'getSupergroup',
        supergroup_id: type.supergroup_id,
      });
      if (supergroup.status?._ !== 'chatMemberStatusCreator') return false;

      const full = await this.client.invoke({
        _: 'getSupergroupFullInfo',
        supergroup_id: type.supergroup_id,
      });
      if (!String(full.description || '').includes(VAULT_MARKER)) return false;
      // Network check to ensure we ACTUALLY have access
      await this.client.invoke({
        _: 'getChatHistory',
        chat_id: chatId,
        from_message_id: 0,
        offset: 0,
        limit: 1,
        only_local: false,
      });
      return true;
    } catch {
      return false;
    }
  }

  async resolveVaultChatId() {
    const cached = this.db.getVaultChatId();
    if (cached != null && (await this.isVaultChat(cached))) return cached;
    const found = await this.findVaultChannel();
    if (found != null) {
      this.db.setVaultChatId(found);
      return found;
    }
    const created = await this.createVaultChannel();
    this.db.setVaultChatId(created);
    return created;
  }

  entryFromMessage(msg) {
    const content = msg.content;
    const id = msg.id;
    if (!content) return null;
    if (content._ === 'messageDocument') {
      const caption = content.caption?.text || '';
      return decodeCaption(id, caption);
    }
    if (content._ === 'messageText') {
      const text = content.text?.text || '';
      return decodeCaption(id, text);
    }
    return null;
  }

  _isDocumentMessage(msg) {
    return msg?.content?._ === 'messageDocument';
  }

  async scanHistory(chatId, onProgress) {
    let fromMessageId = 0;
    let scanned = 0;
    let maxId = this.db.getLastMessageId();
    const seenIds = new Set();

    while (true) {
      const page = await this.client.invoke({
        _: 'getChatHistory',
        chat_id: chatId,
        from_message_id: fromMessageId,
        offset: 0,
        limit: 100,
        only_local: false,
      });
      const messages = page.messages || [];
      if (messages.length === 0) break;

      for (const msg of messages) {
        const id = msg.id;
        seenIds.add(id);
        const entry = this.entryFromMessage(msg);
        if (entry) this.db.upsert(entry);
        scanned += 1;
        if (id > maxId) maxId = id;
      }

      const nextFrom = messages[messages.length - 1].id;
      if (nextFrom === fromMessageId) break;
      fromMessageId = nextFrom;
      onProgress?.(scanned);
    }

    this.db.reconcileToMessageIds(seenIds);
    this.db.setLastMessageId(maxId);
    this._emitChange();
    return scanned;
  }

  /** Re-index folder marker text messages so caption tags match local DB. */
  async resyncDirMarkers(chatId) {
    let fromMessageId = 0;
    while (true) {
      const page = await this.client.invoke({
        _: 'getChatHistory',
        chat_id: chatId,
        from_message_id: fromMessageId,
        offset: 0,
        limit: 100,
        only_local: false,
      });
      const messages = page.messages || [];
      if (messages.length === 0) break;

      for (const msg of messages) {
        if (msg.content?._ !== 'messageText') continue;
        const entry = this.entryFromMessage(msg);
        if (entry?.path.endsWith('/')) this.db.upsert(entry);
      }

      const nextFrom = messages[messages.length - 1].id;
      if (nextFrom === fromMessageId) break;
      fromMessageId = nextFrom;
    }

    this.db.reconcileFolderTagsFromMarkers();
    this._emitChange();
  }

  listenUpdates(chatId) {
    if (this._updateHandler) {
      this.client.off('update', this._updateHandler);
    }
    this._updateHandler = async (u) => {
      try {
        switch (u._) {
          case 'updateNewMessage': {
            const msg = u.message;
            if (msg.chat_id !== chatId) return;
            // Outgoing messages get a temporary negative id until Telegram confirms send.
            if (this._tdId(msg.id) < 0) break;
            if (!this._shouldIndexMessage(msg)) break;
            const entry = this.entryFromMessage(msg);
            if (entry) {
              this.db.upsert(entry);
              this._emitChange();
            }
            break;
          }
          case 'updateMessageSendSucceeded': {
            const msg = u.message;
            if (!msg || msg.chat_id !== chatId) break;
            const oldId = this._tdId(u.old_message_id);
            const newId = this._tdId(msg.id);
            this.rekeyUploadPending(oldId, newId);
            if (oldId !== newId) this.db.rekeyMessageId(oldId, newId);
            if (!this._isDocumentMessage(msg)) {
              const entry = this.entryFromMessage(msg);
              if (entry) {
                this.db.upsert(entry);
                this._emitChange();
              }
            }
            this._onMessageSendSucceeded?.(oldId, newId);
            break;
          }
          case 'updateMessageSendFailed': {
            const oldId = this._tdId(u.old_message_id);
            if (oldId < 0) {
              this.db.delete(oldId);
              this._emitChange();
            }
            this._onMessageSendFailed?.(oldId);
            break;
          }
          case 'updateDeleteMessages': {
            if (u.chat_id !== chatId || !u.is_permanent) return;
            for (const id of u.message_ids || []) {
              this.db.delete(id);
            }
            this._emitChange();
            break;
          }
          case 'updateMessageContent': {
            if (u.chat_id !== chatId) return;
            const msgId = this._tdId(u.message_id);
            if (msgId > 0 && this._pendingUploadIds.has(msgId)) return;
            const entry = this.entryFromMessage({
              id: u.message_id,
              content: u.new_content,
            });
            if (entry) {
              this.db.upsert(entry);
              this._emitChange();
            }
            break;
          }
          default:
            break;
        }
      } catch (e) {
        console.error('[ChannelService update]', e);
      }
    };
    this.client.on('update', this._updateHandler);
  }

  dispose() {
    if (this._updateHandler) {
      this.client.off('update', this._updateHandler);
      this._updateHandler = null;
    }
  }

  /** @param {Record<string, unknown>} msg */
  tdFileIdFromMessage(msg) {
    const content = msg.content;
    if (!content || content._ !== 'messageDocument') return null;
    const doc = content.document;
    if (!doc) return null;
    const inner = doc.document ?? doc;
    const id = inner?.id ?? doc.id;
    if (id == null) return null;
    return typeof id === 'number' ? id : Number.parseInt(String(id), 10);
  }

  /**
   * Scan chatListMain for all channels with #televault-v1, excluding a given chatId.
   * @param {import('tdl').Client} client
   * @param {number|null} excludeChatId - own vault chatId to exclude
   * @returns {Promise<Array<{ chatId: number, title: string }>>}
   */
  static async findAllVaultChannels(client, excludeChatId) {
    const chats = await client.invoke({
      _: 'getChats',
      chat_list: { _: 'chatListMain' },
      limit: 200,
    });
    const results = [];
    for (const chatId of chats.chat_ids || []) {
      if (chatId === excludeChatId) continue;
      try {
        const chat = await client.invoke({ _: 'getChat', chat_id: chatId });
        const type = chat.type;
        if (type?._ !== 'chatTypeSupergroup' || !type.is_channel) continue;
        const full = await client.invoke({
          _: 'getSupergroupFullInfo',
          supergroup_id: type.supergroup_id,
        });
        const description = String(full.description || '');
        const title = String(chat.title || '');
        console.log(`[TeleVault] Checking channel: ${title} (${chatId}) - Description: ${description}`);
        if (description.includes(VAULT_MARKER) || title.includes(VAULT_MARKER)) {
          console.log(`[TeleVault] Found Shared Vault: ${title} (${chatId})`);
          results.push({ chatId, title: chat.title || `Vault ${chatId}` });
        }
      } catch (err) {
        // skip inaccessible chats
        console.log(`[TeleVault] Skipping chat ${chatId} due to error: ${err.message}`);
      }
    }
    return results;
  }
}

module.exports = { ChannelService };
